core/crypto/src/key_file.rs

```diff
@@ -22,10 +57,242 @@ impl KeyFile {
         let mut file = File::open(path).expect("Could not open key file.");
         let mut content = String::new();
         file.read_to_string(&mut content).expect("Could not read from key file.");
         serde_json::from_str(&content).expect("Failed to deserialize KeyFile")
     }
+    
+    // decrypts a file and return an instance of the file
+    pub fn from_encrypted_file(path: &Path, key : &Key) -> Self {
+        let key_file = KeyFile::from_file(path);
+        let ssb = SodiumoxideSecretBox::new();
+        let decrypted_secretkey = ssb.decrypt(&secretKeyToVec(key_file.secret_key), &key);
+        KeyFile {
+            account_id: key_file.account_id,
+            public_key: key_file.public_key,
+            secret_key: secretKeyFromVec(decrypted_secretkey),
+            key_hash: KeyHash::default(),
+        }
+    }
+
+    // if a private key is encrypted , ask for a password in console
+    // if an entered password matches a password entered in step 3, decode an encrypted password otherwise keep asking for a correct password
+    pub fn from_encrypted_file_with_password(path : &Path) -> Result<Self, Error>  {  
+        let mut key_file = KeyFile::from_file(path);
+        if key_file.key_hash.eq(&KeyHash::default()) {
+            let password_input = PasswordInput::process()?; // ask for a password in command line
+            if  password_input.not_empty() { // password is not empty decode a file with a passwordm if password empty do nothing
+                let decryption_keyhash = KeyHash::new(&password_input.content);
+                if decryption_keyhash.eq(&key_file.key_hash) { // if hashes of passwords match then decrypt Keyfile.secret_key
+                    let decryption_key = password_input.gen_key_from_password_input()?;
+                    key_file = KeyFile::from_encrypted_file(path, &decryption_key);
+                }
+            }
+        }
+
+        Ok(key_file)
+    }
+}
+
+pub fn perform_encryption(
+    key_file : &mut KeyFile,
+    key_store_path: &Path,
+    home_dir: &Path,
+) -> Result<(), Error> {
+    key_file.write_to_file(&key_store_path);
+    
+    if home_dir.join("config.json").exists() {
+        //add password corrrectness validation to from_encrypted_file_with_password
+        KeyFile::from_encrypted_file_with_password(&key_store_path)?;
+    } else {
+        key_file.write_to_encrypted_file(&key_store_path)?;
+    }
+
+    Ok(())
+}
+
+pub struct PasswordInput {
+    pub content: Vec<u8>,
+    pub length: usize
+}
+
+impl PasswordInput {
+    // should support that user might not want to encrypt the key
+    pub fn enter_from_console() -> Self {
+        println!("Enter a password for a secret key: ");
+        let mut input = String::new();
+        let _ = std::io::stdin().read_line(&mut input).expect("Could not read from stdin");
+        let content = input.trim().as_bytes().to_vec();
+        let length = input.len();
+        PasswordInput { content, length }
```

Comment:
I don't think there is good reason to have it into multiple functions. Just do everything in `enter_from_console` or if you need to make it injectable for testing split it into two functions.